ui/StatusQ/src/movablemodel.cpp

```diff
@@ -0,0 +1,280 @@
+}
+
+int MovableModel::rowCount(const QModelIndex &parent) const
+{
+    if (m_detached)
+        return m_indexes.size();
+
+    if (m_sourceModel == nullptr)
+        return 0;
+
+    return m_sourceModel->rowCount();
+}
+
+QVariant MovableModel::data(const QModelIndex &index, int role) const
+{
+    if (!checkIndex(index, CheckIndexOption::IndexIsValid))
+        return {};
+
+    if (m_detached)
+        return m_indexes.at(index.row()).data(role);
+
+    if (m_sourceModel == nullptr)
+        return {};
+
+    return m_sourceModel->data(index, role);
+}
+
+QHash<int, QByteArray> MovableModel::roleNames() const
+{
+    if (m_sourceModel == nullptr)
+        return {};
+
+    return m_sourceModel->roleNames();
+}
+
+void MovableModel::detach()
+{
+    if (m_detached || m_sourceModel == nullptr)
+        return;
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsAboutToBeInserted, this,
+               &MovableModel::beginInsertRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsInserted, this,
+               &MovableModel::endInsertRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsAboutToBeRemoved, this,
+               &MovableModel::beginRemoveRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsRemoved, this,
+               &MovableModel::endRemoveRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsAboutToBeMoved, this,
+               &MovableModel::beginMoveRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::rowsMoved, this,
+               &MovableModel::endMoveRows);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::dataChanged, this,
+               &MovableModel::dataChanged);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::layoutAboutToBeChanged, this,
+               &MovableModel::layoutAboutToBeChanged);
+
+    disconnect(m_sourceModel, &QAbstractItemModel::layoutChanged, this,
+               &MovableModel::layoutChanged);
+
+    connect(m_sourceModel, &QAbstractItemModel::dataChanged, this,
+            [this](const QModelIndex& topLeft, const QModelIndex& bottomRight,
+            const QVector<int>& roles) {
+        emit dataChanged(index(0), index(rowCount() - 1), roles);
+    });
+
+    connect(m_sourceModel, &QAbstractItemModel::rowsInserted, this,
+            [this](const QModelIndex &parent, int first, int last) {
+
+        beginInsertRows({}, first, last);
+
+        int oldCount = m_indexes.size();
+        int insertCount = last - first + 1;
+
+        m_indexes.reserve(m_indexes.size() + insertCount);
+
+        for (auto i = first; i <= last; i++)
+            m_indexes.emplace_back(m_sourceModel->index(i, 0));
+
+        std::rotate(m_indexes.begin() + first, m_indexes.begin() + oldCount,
+                    m_indexes.end());
+
+        endInsertRows();
+    });
+
+    connect(m_sourceModel, &QAbstractItemModel::rowsAboutToBeRemoved, this,
+            [this] (const QModelIndex &parent, int first, int last) {
```

Comment:
Consider wrapping this in a separate method for better readability.

Reply:
Yeah, moving to another functions would make this one shorter. But having all handler in one place is also to some degree beneficial. I would leave it as it is for now.